### DISCUSSION TITLE
python-build fails to build Python 3.11.0 against brew-built openssl

### DISCUSSION THREAD

**Discussion by jacktose** [Homebrew/discussions#3898] [Linux] [answered]
I am failing to build python 3.10 and 3.11 against brewed `openssl@1.1` and `@3`, using brewed `gcc`, `glibc`, `ld`, etc.
I can see that `ld` throws errors like: ```.../libcrypto.so: undefined reference to `dladdr@GLIBC_2.34'``` but I can't find the root cause.

I'm on an unsupported config and mixing brewed stuff with outside-of-brew stuff, so I know this is no Issue. Just looking for any tips from someone who understands compilation better than I do—that is, understands it at all.

## Context
[Here's my brew situation][01_brew-context]: the output of `brew config`, `list --versions`, & `doctor`. (Yes, doctor is messy. I don't control this machine.)
Note `gcc 12.2.0` and `glibc 2.35_1`.

## Problem
python-build (& pyenv, asdf) fails to build python because of something about OpenSSL:
```
$ CONFIGURE_OPTS="--with-openssl=$(brew --prefix openssl@3) --with-openssl-rpath=auto" python-build 3.11.0 ~/python-build/3.11.0/
Downloading Python-3.11.0.tar.xz...
-> https://www.python.org/ftp/python/3.11.0/Python-3.11.0.tar.xz
Installing Python-3.11.0...
ERROR: The Python ssl extension was not compiled. Missing the OpenSSL lib?

Please consult to the Wiki page to fix the problem.
https://github.com/pyenv/pyenv/wiki/Common-build-problems

BUILD FAILED (CentOS 7.9.2009 using python-build 20180424)

Inspect or clean up the working tree at /tmp/python-build.20221108142742.23728
Results logged to /tmp/python-build.20221108142742.23728.log
```
[(full output here)][02_python-build-output]
I have experimented with the suggestions in the [wiki], with no change. This invocation worked with `openssl@1.1` and python `3.10.8` a few weeks ago. Now I get effectively these same results with any combination of those versions. All that has changed to my knowledge is updating a few brewed formula/bottles, including `openssl@1.1` from `1.1.1q` or `r` to `s`.

## Logs
[Here's python-build's log.][03_python-build.20221108135045.31869.log] Note the following errors (search `ssl` for more):
```
checking for openssl/ssl.h in /data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3... yes
checking whether compiling and linking against OpenSSL works... no
checking for --with-openssl-rpath... auto
checking whether OpenSSL provides required ssl module APIs... no
checking whether OpenSSL provides required hashlib module APIs... no
```
That clearly shows _that_ it failed, but not much more. So I look at the real [`config.log`][04_config.log], from the working directory:
<details>
<summary><code>config.log</code> excerpt</summary>

```
configure:22831: checking for openssl/ssl.h in /data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3
configure:22838: result: yes
configure:22854: checking whether compiling and linking against OpenSSL works
Trying link with OPENSSL_LDFLAGS=-L/data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/lib; OPENSSL_LIBS=-lssl -lcrypto; OPENSSL_INCLUDES=-I/data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/include
configure:22876: gcc -pthread -o conftest  -I/data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/include -I/usr/lusers/jacktose/python-build/3.11.0//include  -L/usr/lusers/jacktose/python-build/3.11.0//lib  -L/data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/lib conftest.c -lssl -lcrypto -lpthread -ldl  -lutil >&5
/usr/lusers/jacktose/.linuxbrew/bin/ld: /data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/lib/libcrypto.so: undefined reference to `dladdr@GLIBC_2.34'
/usr/lusers/jacktose/.linuxbrew/bin/ld: /data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/lib/libcrypto.so: undefined reference to `dlclose@GLIBC_2.34'
/usr/lusers/jacktose/.linuxbrew/bin/ld: /data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/lib/libcrypto.so: undefined reference to `fstat@GLIBC_2.33'
/usr/lusers/jacktose/.linuxbrew/bin/ld: /data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/lib/libcrypto.so: undefined reference to `dlerror@GLIBC_2.34'
/usr/lusers/jacktose/.linuxbrew/bin/ld: /data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/lib/libcrypto.so: undefined reference to `pthread_key_create@GLIBC_2.34'
/usr/lusers/jacktose/.linuxbrew/bin/ld: /data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/lib/libcrypto.so: undefined reference to `pthread_rwlock_unlock@GLIBC_2.34'
/usr/lusers/jacktose/.linuxbrew/bin/ld: /data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/lib/libcrypto.so: undefined reference to `pthread_setspecific@GLIBC_2.34'
/usr/lusers/jacktose/.linuxbrew/bin/ld: /data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/lib/libcrypto.so: undefined reference to `pthread_rwlock_destroy@GLIBC_2.34'
/usr/lusers/jacktose/.linuxbrew/bin/ld: /data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/lib/libcrypto.so: undefined reference to `pthread_rwlock_wrlock@GLIBC_2.34'
/usr/lusers/jacktose/.linuxbrew/bin/ld: /data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/lib/libcrypto.so: undefined reference to `dlsym@GLIBC_2.34'
/usr/lusers/jacktose/.linuxbrew/bin/ld: /data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/lib/libcrypto.so: undefined reference to `pthread_getspecific@GLIBC_2.34'
/usr/lusers/jacktose/.linuxbrew/bin/ld: /data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/lib/libcrypto.so: undefined reference to `pthread_key_delete@GLIBC_2.34'
/usr/lusers/jacktose/.linuxbrew/bin/ld: /data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/lib/libcrypto.so: undefined reference to `pthread_rwlock_init@GLIBC_2.34'
/usr/lusers/jacktose/.linuxbrew/bin/ld: /data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/lib/libcrypto.so: undefined reference to `stat@GLIBC_2.33'
/usr/lusers/jacktose/.linuxbrew/bin/ld: /data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/lib/libcrypto.so: undefined reference to `dlopen@GLIBC_2.34'
/usr/lusers/jacktose/.linuxbrew/bin/ld: /data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/lib/libcrypto.so: undefined reference to `pthread_once@GLIBC_2.34'
/usr/lusers/jacktose/.linuxbrew/bin/ld: /data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/lib/libcrypto.so: undefined reference to `pthread_rwlock_rdlock@GLIBC_2.34'
collect2: error: ld returned 1 exit status
configure:22876: $? = 1
configure: failed program was:
| /* confdefs.h */
| #define _NETBSD_SOURCE 1
```
</details>

So I copy out the failed program to try just that part.

## Conftest
Here's [`conftest.c`][05_conftest.c]. The main event seems to be: `#include <openssl/ssl.h>`.
I invoke `gcc` on it as per `config.log` and get the same errors:
```
$ gcc -pthread -o conftest  -I/data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/include -I/usr/lusers/jacktose/python-build/3.11.0//include  -L/usr/lusers/jacktose/python-build/3.11.0//lib  -L/data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/lib conftest.c -lssl -lcrypto -lpthread -ldl  -lutil
/usr/lusers/jacktose/.linuxbrew/bin/ld: /data/shared/lusers/jacktose/.linuxbrew/opt/openssl@3/lib/libcrypto.so: undefined reference to `dladdr@GLIBC_2.34'
[16 more like that]
collect2: error: ld returned 1 exit status
```
[(full output here)][06_conftest-result]

## What now?
So there's my minimum reproducible problem so far, but now I'm tapped out (no homebrew pun intended). Looking for stuff `@GLIBC_2.34` seems wrong when I have brewed glibc 2.<em>35</em>_1, but I have no idea how to go deeper toward a root cause. Help?

[gist]: https://gist.github.com/jacktose/7d11ae3800153c5fc95ee76482a96870
[00_info.txt]: https://gist.githubusercontent.com/jacktose/7d11ae3800153c5fc95ee76482a96870/raw/dfa4f790eeecf6df0b10bd7757ff4152f0e27a3e/00_info.txt
[01_brew-context]: https://gist.githubusercontent.com/jacktose/7d11ae3800153c5fc95ee76482a96870/raw/dfa4f790eeecf6df0b10bd7757ff4152f0e27a3e/01_brew-context
[02_python-build-output]: https://gist.githubusercontent.com/jacktose/7d11ae3800153c5fc95ee76482a96870/raw/dfa4f790eeecf6df0b10bd7757ff4152f0e27a3e/02_python-build-output
[03_python-build.20221108135045.31869.log]: https://gist.githubusercontent.com/jacktose/7d11ae3800153c5fc95ee76482a96870/raw/dfa4f790eeecf6df0b10bd7757ff4152f0e27a3e/03_python-build.20221108135045.31869.log
[04_config.log]: https://gist.githubusercontent.com/jacktose/7d11ae3800153c5fc95ee76482a96870/raw/dfa4f790eeecf6df0b10bd7757ff4152f0e27a3e/04_config.log
[05_conftest.c]: https://gist.githubusercontent.com/jacktose/7d11ae3800153c5fc95ee76482a96870/raw/dfa4f790eeecf6df0b10bd7757ff4152f0e27a3e/05_conftest.c
[06_conftest-result]: https://gist.githubusercontent.com/jacktose/7d11ae3800153c5fc95ee76482a96870/raw/dfa4f790eeecf6df0b10bd7757ff4152f0e27a3e/06_conftest-result

[wiki]: https://github.com/pyenv/pyenv/wiki/Common-build-problems#error-the-python-ssl-extension-was-not-compiled-missing-the-openssl-lib

**Comment by Bo98** [comment, marked as answer]
Which GCC are you using? If you are using Ubuntu < 22.04 (i.e, you are using brewed glibc) and are using Homebrew installed libraries, then you must also use Homebrew GCC (`gcc-12`).

System GCC may work but you'll need to provide some glibc flags if you use that.

**Reply by jacktose** [reply]
Hey, that's it! At least, specifying brew's `gcc-12` gets me past openssl to a different error.
```
$ CC="/usr/lusers/jacktose/.linuxbrew/bin/gcc-12" \
CONFIGURE_OPTS="--with-openssl=/usr/lusers/jacktose/.linuxbrew/opt/openssl@3 --with-openssl-rpath=auto" \
python-build -v 3.11.0 ~/python-build/3.11.0
...
WARNING: The Python sqlite3 extension was not compiled. Missing the SQLite3 lib?
Installed Python-3.11.0 to /usr/lusers/jacktose/python-build/3.11.0

$ ~/python-build/3.11.0/bin/python
Python 3.11.0 (main, Nov  9 2022, 21:31:28) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ssl
>>> import sqlite3
...
ModuleNotFoundError: No module named '_sqlite3'
```
(I'm on CentOS 7, for my sins.)
